### PR TITLE
feat(display-assets): display nested assets in model

### PR DESCRIPTION
**Pull request by AaronCGoidel**

### CODE REVIEW

**Comment by duksis** [comment]
@dazlious @lmtejada This PR adds functionality for retrieving/listing assets attached to a DatoCMS record.
your feedback and improvement suggestions would be highly appreciated

File: src/components/Main/Main.jsx [hunk 1]
```diff
@@ -1,15 +1,202 @@
...
+    let ids = [];
+    if (field.field_type === 'file') {
+      ids.push(plugin.getFieldValue(field.api_key).upload_id);
+    } else {
+      // this is a link to a model instance
+      let itemId = plugin.getFieldValue(field.api_key);
+      ids = ids.concat(await parse_nested_content(await client.items.find(itemId)));
+    }
+    return ids;
+  };
```
Comment (dazlious) [inline src/components/Main/Main.jsx:72]: Why create array here? 

```suggestion
const parse_field = async ({ field_type: fieldType, api_key: apiKey }) => {
  switch(fieldType) {
    case 'file':
      return plugin.getFieldValue(field.api_key).upload_id;
    case default:
      const itemId = plugin.getFieldValue(apiKey);
      return await parse_nested_content(await client.items.find(itemId))
  }
}
```

File: src/components/Main/Main.jsx [hunk 2]
```diff
@@ -1,15 +1,202 @@
-import React from 'react';
+import React, { useEffect, useState } from 'react';
 import { bool, shape } from 'prop-types';
 
 import connectToDatoCms from '@root/hoc/connectToDatoCms';
 
 import './Main.scss';
+import { map } from 'lodash-es';
```
Comment (dazlious) [inline src/components/Main/Main.jsx:7]: js imports should come before scss

File: src/components/Main/Main.jsx [hunk 3]
```diff
@@ -1,15 +1,202 @@
...
 
 const Main = ({ client, devMode, plugin }) => {
   if (devMode) {
     console.debug({ client, plugin }); // eslint-disable-line no-console
   }
-  return <div className="container">Your Plugin goes here</div>;
+
+  let [source, setSource] = useState('');
```
Comment (dazlious) [inline src/components/Main/Main.jsx:14]: always go with const. You will see when you will need a let. But usually 99% of the time it is const in JSX.

File: src/components/Main/Main.jsx [hunk 4]
```diff
@@ -1,15 +1,202 @@
...
+  let [doneLoading, setDoneLoading] = useState(false);
+
+  const get_model = async (id) => {
+    try {
+      let ret = await client.items.find(id, {});
+      return ret;
+    } catch (err) {
+      return {};
+    }
+  };
```
Comment (dazlious) [inline src/components/Main/Main.jsx:28]:
```suggestion
  const getModel = async (id) => {
    let model = null;
    try {
      model = await client.items.find(id, {});
    } catch (err) {
      if (devMode) {
        console.error(`Could not find model with id ${id}`);
      }
    }
    return model;
  };
```

File: src/components/Main/Main.jsx [hunk 5]
```diff
@@ -1,15 +1,202 @@
...
+
+  const parse_nested_content = async (obj) => {
+    let ids = [];
```
Comment (dazlious) [inline src/components/Main/Main.jsx:31]: const ids. You do not need to reassign ids. Just push onto it

File: src/components/Main/Main.jsx [hunk 6]
```diff
@@ -1,15 +1,202 @@
...
+    for (const property in obj) {
+      if (typeof obj[property] === 'string' && property !== 'id') {
```
Comment (dazlious) [inline src/components/Main/Main.jsx:33]: do `const field = obj[property];`

File: src/components/Main/Main.jsx [hunk 7]
```diff
@@ -1,15 +1,202 @@
...
+        // this could be a linked instance
+        let nested = await get_model(obj[property]);
+
+        if (nested) {
+          ids = ids.concat(await parse_nested_content(nested));
```
Comment (dazlious) [inline src/components/Main/Main.jsx:38]: push onto ids, so you do not need the let

**Review by dazlious — CHANGES_REQUESTED** [review]
LGTM overall, some smaller improvements can be made to improve the code quality and readability. Please review the loops with async/await and make sure they really await. Easy way to test that is to run a console log within the loop that iterates and a done log after the loop. If the done is fired before the iteree fire, something is not working correctly.

Some best practices for JS and specifically React:
- always use const and make everything immutable
- use camelCase
- do not repeat yourself

**Pull request merged**